src/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs

```diff
@@ -429,6 +429,80 @@ public static bool TrySkipBlock(BinaryReader reader)
 
             return true;
         }
+
+        public static bool TryValidateBlock(BinaryReader reader, ZipArchiveEntry entry)
+        {
+            const int OffsetToFilenameLength = 22; // from the point after the signature
+
+            if (reader.ReadUInt32() != SignatureConstant)
+                return false;
+
+            if (reader.BaseStream.Length < reader.BaseStream.Position + OffsetToFilenameLength)
+                return false;
+
+            uint minimunVersion = reader.ReadUInt16();
+            int dataDescriptorBit = reader.ReadUInt16() & 0x08;
+            reader.BaseStream.Seek(10, SeekOrigin.Current);
+            long compressedSize = reader.ReadUInt32();
+            long uncompressedSize = reader.ReadUInt32();
+            int filenameLength = reader.ReadUInt16();
+            uint extraFieldLength = reader.ReadUInt16();
+            reader.BaseStream.Seek(filenameLength, SeekOrigin.Current);
+            long endExtraFields = reader.BaseStream.Position + extraFieldLength;
+            
+            if (dataDescriptorBit == 0)
+            {
+                bool uncompressedSizeInZip64 = uncompressedSize == ZipHelper.Mask32Bit;
+                bool compressedSizeInZip64 = compressedSize == ZipHelper.Mask32Bit;
+                Zip64ExtraField zip64;
+
+                // Ideally we should also check if the minimunVersion is 64 bit or above, but there could zip files for which this version is not set correctly
+                if (uncompressedSizeInZip64 || compressedSizeInZip64)
+                {
+                    zip64 = Zip64ExtraField.GetJustZip64Block(new SubReadStream(reader.BaseStream, reader.BaseStream.Position, extraFieldLength), uncompressedSizeInZip64, compressedSizeInZip64, false, false);
+
+                    if (zip64.UncompressedSize != null)
+                        uncompressedSize = zip64.UncompressedSize.Value;
+                    if (zip64.CompressedSize != null)
+                        compressedSize = zip64.CompressedSize.Value;
+                }
+
+                reader.BaseStream.AdvanceToPosition(endExtraFields);
+            }
+            else
+            {
+                bool is64bit = minimunVersion > 44;
+                long seekSize = entry.CompressedLength + (is64bit ? 8 : 4) * 2;
+                uint dataDescriptorSignature = reader.ReadUInt32();
```

Comment:
I don't believe that this is the correct place to read an optional descriptor signature.  If I look at our code that writes it, it's writing immediately before the CRC in the descriptor itself.  So I think you need to change the code at 484 to account for the optional descriptor there preceeding the CRC.  See https://github.com/dotnet/corefx/blob/0eed4042f056b4d9f6f4d1bf2918e234d67a6f88/src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs#L1023-L1045.

Reply:
Good catch, thanks